src/transformers/models/deepseek_v3/modeling_deepseek_v3.py

```diff
@@ -168,25 +168,44 @@ def __init__(self, config):
         )
 
     def moe(self, hidden_states: torch.Tensor, topk_indices: torch.Tensor, topk_weights: torch.Tensor):
-        r"""
-        CALL FOR CONTRIBUTION! I don't have time to optimise this right now, but expert weights need to be fused
-        to not have to do a loop here (deepseek has 256 experts soooo yeah).
         """
-        final_hidden_states = torch.zeros_like(hidden_states, dtype=topk_weights.dtype)
-        expert_mask = torch.nn.functional.one_hot(topk_indices, num_classes=len(self.experts))
-        expert_mask = expert_mask.permute(2, 0, 1)
-
-        for expert_idx in range(len(self.experts)):
-            expert = self.experts[expert_idx]
-            mask = expert_mask[expert_idx]
-            token_indices, weight_indices = torch.where(mask)
-
-            if token_indices.numel() > 0:
-                expert_weights = topk_weights[token_indices, weight_indices]
-                expert_input = hidden_states[token_indices]
-                expert_output = expert(expert_input)
-                weighted_output = expert_output * expert_weights.unsqueeze(-1)
-                final_hidden_states.index_add_(0, token_indices, weighted_output)
+        Fully vectorized MoE (no Python loop).
+        Assumes all experts are the same class and can be stacked.
+        """
+        num_tokens, hidden_dim = hidden_states.shape
+        top_k = topk_indices.shape[1]
+        num_experts = len(self.experts)
+
+        # Stack expert parameters
+        # (num_experts, hidden_dim, intermediate_dim)
+        w1 = torch.stack([e.gate_proj.weight for e in self.experts], dim=0)
+        w2 = torch.stack([e.down_proj.weight for e in self.experts], dim=0)
+        w3 = torch.stack([e.up_proj.weight for e in self.experts], dim=0)
```

Comment:
there is no way this is not super costly no? Also have you ran this with TP? we need to make sure both pathes work!